mautic/mautic
# Feature - Add custom_args for Sendgrid API

| Q                                      | A
| -------------------------------------- | ---
| Bug fix?           | Not really
| New feature/enhancement?       | Yes
| Deprecations?                          | No
| BC breaks? (use the c.x branch)        | No
| Automated tests included?              | TODO (waiting for tests results)
| Related user documentation PR URL      | I cloned to add, but "Configuration" is not documented anywhere (e-mail settings are only cited on 03.Setup, but I dint change anything there). The messages.ini contains some explanation of this field for the user.



#### Description:

This feature enables system-admins to set custom_args to be set on all emails sent by Sendgrid API transport.

This can be configured on the Email Settings page.




#### Steps to test this PR:


1. Load up [this PR](https://mautibox.com)
2. Make sure to use SendGrid API and not Sendgrid SMTP as a transport method in *Mautic Configuration* -> *Email* => on *Mail Send Settings* set "Service to send mail through" as SendGrid API ![image](https://user-images.githubusercontent.com/1919076/146460825-e8577dc3-e5bf-48b9-b07b-f5460647df41.png)
2. Still in the configuration page, add some key/value in the field "Custom Arguments"
3. Start ngrok or tool to monitor Sendgrid Webhook request
4. Set up sendgrid webhook to your address (Login to https://app.sendgrid.net => on left menu, *Mail Settings* -> *Mail Settings* => *Event Webhook* => type your webserver hook URL, select some all events or at least the "processed" event and click save.
5. Create an contact with e-amil field, send an e-mail via contact page or email campaign
6. Watch ngrok/webserver for the callback, the request should include the setup Custom Arguments set on step 2 ![image](https://user-images.githubusercontent.com/1919076/146462074-ada2ea1a-6c3d-40bb-8aba-5e67f4410f04.png)


**This code is backward compatible** and do not change the current behavior if not configured, even if the `mailer_transport_custom_args` is missing from `app/config/local.php`, it only appears after configuration is saved.


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10691"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

This feature somewhat make useless the "always-keep-custom-header" setting I commented on #10688  

If you are curious, use case is more described better in https://github.com/AppCivico/sendgrid-fanout#why
